### PR TITLE
Feature/7.x/virtual connection

### PR DESCRIPTION
**Pull request by Mpdreamz**
Moves our virtual clustering we use in tests into its own package.

### CODE REVIEW

**Comment by Mpdreamz** [comment]
Rebased, the idea behind spinning this off into its own project is

To help testing, since many interfaces are gone mocking is not an option. This is by design we feel that you should mock the boundaries you control in your application (service layers/etc) injecting a connection that behaves a certain way can aid with testing though and `VirtualConnection` can help out here.

E.g for https://github.com/elastic/apm-agent-dotnet/pull/329 I want to use this to write unit test for the clients integration with APM.

File: src/Auxiliary/Elasticsearch.Net.Virtual/VirtualizedCluster.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Elasticsearch.Net.Virtual.Providers;
+
+namespace Elasticsearch.Net.Virtual
+{
+	public class VirtualizedCluster
```
Comment (russcam) [inline src/Auxiliary/Elasticsearch.Net.Virtual/VirtualizedCluster.cs]: `VirtualCluster`, to align with `VirtualConnection`?

File: build/Elasticsearch.Net.Virtual.nuspec [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
+	<metadata>
+		<id>Elasticsearch.Net.Virtual</id>
```
Comment (russcam) [inline build/Elasticsearch.Net.Virtual.nuspec]: Naming is hard 😃 

`Elasticsearch.Net.Virtual` feels somewhat vague. The package is a way to mock/stub a cluster, with a specific `IConnection` implementation. Would `Elasticsearch.Net.VirtualCluster`, or  `Elasticsearch.Net.MockCluster` or `Elasticsearch.Net.StubCluster` or `Elasticsearch.Net.FakeCluster` better convey the usage?

File: build/Elasticsearch.Net.Virtual.nuspec [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		<version>$version$</version>
+		<title>Elasticsearch.Net.Virtual - A highly configurable in memory IConnection</title>
+		<authors>Elastic and contributors</authors>
+		<owners>Elastic</owners>
+		<license type="file">license.txt</license>
+		<projectUrl>https://github.com/elastic/elasticsearch-net</projectUrl>
+		<iconUrl>https://raw.githubusercontent.com/elastic/elasticsearch-net/master/build/nuget-icon.png</iconUrl>
+		<requireLicenseAcceptance>false</requireLicenseAcceptance>
+		<summary>Provides a way to assert cluent behaviour through a rule engine backed VirtualClusterConnection</summary>
```
Comment (russcam) [inline build/Elasticsearch.Net.Virtual.nuspec]:
```suggestion
		<summary>Provides a way to assert client behaviour through a rule engine backed VirtualClusterConnection</summary>
```

File: build/Elasticsearch.Net.Virtual.nuspec [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		<description>Provides a way to assert cluent behaviour through a rule engine backed VirtualClusterConnection</description>
```
Comment (russcam) [inline build/Elasticsearch.Net.Virtual.nuspec]:
```suggestion
		<description>Provides a way to assert client behaviour through a rule engine backed VirtualClusterConnection</description>
```

File: build/Elasticsearch.Net.Virtual.nuspec [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+		<releaseNotes>See https://github.com/elastic/elasticsearch-net/releases/tag/$version$</releaseNotes>
+		<copyright>2014-$year$ Elasticsearch BV</copyright>
```
Comment (russcam) [inline build/Elasticsearch.Net.Virtual.nuspec]:
```suggestion
		<copyright>2019-$year$ Elasticsearch BV</copyright>
```

File: build/Elasticsearch.Net.Virtual.nuspec [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		<tags>elasticsearch,elastic,search,lucene,nest</tags>
```
Comment (russcam) [inline build/Elasticsearch.Net.Virtual.nuspec]:
```suggestion
		<tags>elasticsearch,elastic,search,lucene,nest,tests,testing</tags>
```

File: build/Elasticsearch.Net.Virtual.nuspec (hunk 4 above)
Comment (russcam) [inline build/Elasticsearch.Net.Virtual.nuspec]: Might be useful to also mention that it's useful for testing purposes?

**Pull request merged**